### ISSUE TITLE
Incorrect gradients in NaN-ignoring MSE

### ISSUE THREAD

**Issue by randolf-scholz** [pytorch/pytorch#89543]
### 🐛 Describe the bug

```python
#!/usr/bin/env python

import torch

N = 16
# create some observations with missing values
y = torch.randn(N)
y = torch.where(y > .5, y, torch.nan)
x = torch.nn.Parameter(torch.randn(N), requires_grad=True)

# variant 1 correctly produces NaN-free gradient
r = x - y
m = torch.isnan(y)
r = torch.where(m, 0.0, r) **2
loss = r.sum() / m.sum()
loss.backward()
assert not any(torch.isnan(x.grad))  # ✔

# variant 2 incorrectly produces NaN-gradients
x.grad = None
r = (x - y) ** 2
m = torch.isnan(y)
r = torch.where(m, 0.0, r)
loss = r.sum() / m.sum()
loss.backward()
assert not any(torch.isnan(x.grad)) # ✘
```

Both variants are mathematically equivalent and should return the same, NaN-free gradient.

### Versions

<details><summary>versions</summary>

```
Collecting environment information...
PyTorch version: 1.13.0+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.1 LTS (x86_64)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.10.6 (main, Nov  2 2022, 18:53:38) [GCC 11.3.0] (64-bit runtime)
Python platform: Linux-5.15.0-53-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3090
Nvidia driver version: 520.61.05
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.6.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy==0.991
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.23.5
[pip3] torch==1.13.0
[pip3] torchmetrics==0.10.3
[conda] No relevant packages
```

</details>

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

**Comment by randolf-scholz** [comment]
I suspect the problem is with the element-wise power functions, because for the mean absolute error, both variants work correctly.

**Comment by jlukehubbard** [comment]
I've tried running this same test program, I get the same outputs and the only difference I can see between the two code variants is the value of grad_fn for the two resultant tensors `r` . In the code snippet that works, the grad_fn is `PowBackward0` and for the snippet that works the grad_fn field is `WhereBackward0`. Could this issue be cause by autograd's handling of the `where` operation?

**Comment by ZhaoqiongZ** [comment]
for the first code snippet the gradient of r will not related to (x-y)  but the gradient second code snippet will related to (x-y) since the gradient of (x-y)**2 is 2*(x-y),  and y contains nan, so I think the result is expected.

**Comment by randolf-scholz** [comment]
@ZhaoqiongZ But that should be ignored. Consider the case when the mask is defined independently of y:

```python
import torch

x = torch.nn.Parameter(torch.tensor([1.1, 1.2, 1.3]))
y = torch.tensor([2, 3, 4])
m = torch.tensor([True, True, False])

x.grad = None
r = (x - y) ** 2
r = torch.where(m, r, 0.0)
loss = r.sum() / m.sum()
loss.backward()
print(x.grad)
assert not any(torch.isnan(x.grad))  # ✔
```

The value of `x.grad[2]` will always be 0, independently of the value of `y[2]`. Except when `y[2]` is one of `float('nan')`, `float('inf')` or `float('-inf')`, then it breaks, likely due to incorrect handling of the exceptional case. I suspect the backward of `where` might be performing a multiplication with 0 instead of a `where` operation.

**Comment by randolf-scholz** [comment]
## Gradient in case 1.

Define intermediate quantities: $r = x - y$, $z = [m \mathbin{?} r : 𝟎]$, $y = ∑ z_i^2$


$$\begin{aligned}
\frac{∂ f(x)}{∂x} 
   &= \frac{∂y}{∂z}∘\frac{∂z}{∂r}∘\frac{∂r}{∂x} \\
 &= [∆z ↦ ⟨2z∣∆z⟩] ∘ [∆r ↦ [m \mathbin{?} ∆r : 𝟎 ]] ∘ [∆x ↦ ∆x]  & g_z &= 2z \\
 &= [∆r ↦ ⟨2z∣[m \mathbin{?} ∆r : 𝟎 ]⟩] ∘ [∆x ↦ ∆x] \\
 \text{(dubious step)} &= [∆r ↦ ⟨[m \mathbin{?}  2z : 𝟎]∣ ∆r⟩]  ∘ [∆x ↦ ∆x]          \\
 &= [∆r ↦ ⟨[m \mathbin{?} 2[m \mathbin{?} x -y : 𝟎] : 𝟎]∣ ∆r⟩]  ∘ [∆x ↦ ∆x]         \\
 &= [∆r ↦ ⟨[m \mathbin{?} 2(x -y) : 𝟎]∣ ∆r⟩]  ∘ [∆x ↦ ∆x]   & g_r &= [m \mathbin{?}  2(x-y) : 𝟎]     \\
 &= [∆x ↦ ⟨[m \mathbin{?}  2(x-y) : 𝟎]∣ ∆x⟩]                         & g_x &= [m \mathbin{?}  2(x-y) : 𝟎]
\end{aligned}$$

## Gradient in case 2.

Define intermediate quantities: $s = (x-y)^{⊙2}$, $z = [m \mathbin{?} s : 𝟎]$, $y = ∑ z_i$

$$\begin{aligned}
\frac{∂ f(x)}{∂x} 
   &= \frac{∂y}{∂z}∘\frac{∂z}{∂s}∘\frac{∂s}{∂x} \\
 &= [∆z ↦ ⟨𝟏∣∆z⟩] ∘ [∆s ↦ [m \mathbin{?} ∆s : 𝟎 ]] ∘ [∆x ↦ 2(x-y)⊙∆x]  & g_z &= 𝟏 \\
 &= [∆s ↦ ⟨𝟏∣[m \mathbin{?} ∆s : 𝟎 ]⟩] ∘ [∆x ↦  2(x-y)⊙∆x] \\
 \text{(dubious step)} &= [∆s ↦ ⟨[m \mathbin{?}  𝟏 : 𝟎]∣ ∆s⟩] ∘ [∆x ↦ 2(x-y)⊙∆x]      & g_s &= [m \mathbin{?}  𝟏 : 𝟎] \\
 &= [∆x ↦ ⟨[m \mathbin{?}  𝟏 : 𝟎]∣ 2(x-y)⊙∆x⟩] \\
 &= [∆x ↦ ⟨[m \mathbin{?}  𝟏 : 𝟎]⊙2(x-y)∣ ∆x⟩] \\
\text{(incorrect)} &= [∆x ↦ ⟨[m \mathbin{?}  𝟏⊙2(x-y) : {\color{red}𝟎⊙2(x-y)]}∣ ∆x⟩]           & g_x &= [m \mathbin{?}  2(x-y) : 𝟎⊙2(x-y)] \\
\text{(correct)} &= [∆x ↦ ⟨[m \mathbin{?}  𝟏⊙2(x-y) : 𝟎]∣ ∆x⟩]           & g_x &= [m \mathbin{?}  2(x-y) : 𝟎]
\end{aligned}$$

------

We can see that performing the gradient calculation naively one would multiply both arguments to the `where` operator with the inner gradient to produce the new outer gradient. This gives an incorrect result because multiplying `𝟎⋅𝙽𝚊𝙽 = 𝙽𝚊𝙽`, however, this is incorrect in this specific case, because the `𝟎` signifies that the output is categorically independent of `x`. (A where operator is always in essence a branching operation.)

## Correct backwad of the `where` operator

I suspect this is due to an error in the definition of the where backward. Thinking about it a bit, I came up with the following formula for the correct backward of the `where` operator:

Let $y=[m \mathbin{?} a : b]$ and $a$ and $b$ functions of $x$. then

$$\begin{aligned}
\frac{∂y}{∂x} &= \frac{∂y}{∂(a, b)} ∘ \frac{∂(a,b)}{∂x} \\ 
&= \Bigl[ (∆a, ∆b) ⟼ [m  \mathbin{?} {\color{green}{[m \mathbin{?} ∆a : 𝟎 ]}} : {\color{green}{[m \mathbin{?} 𝟎 : ∆b ]}}\Bigr]∘[∆x ⟼ \bigl(\tfrac{∂a}{∂x}{∆x}， \tfrac{∂b}{∂x}{∆x}\bigr) ] \\
\text{(correct)}&= \Bigl[ ∆x  ⟼ [m  \mathbin{?} {\color{green}{[m \mathbin{?} \tfrac{∂a}{∂x}{∆x} : 𝟎 ]}} : {\color{green}{[m \mathbin{?} 𝟎 : \tfrac{∂b}{∂x}{∆x} ]}}\Bigr] \\
\text{(incorrect)}&= \Bigl[ ∆x  ⟼ [m  \mathbin{?} {\color{red}{\tfrac{∂a}{∂x}{∆x}}} : {\color{red}{\tfrac{∂b}{∂x}{∆x}}}\Bigr] 
\end{aligned}$$



This makes things much more complicated. Does torch allow checking whether the outputs are connected to the inputs in the graph? Because the mathematically correct backward for the where operator should be, assuming m is independent of x:

$$\frac{∂}{∂x} [m\mathbin{?} a: b] =  [m\mathbin{?}  \frac{∂a}{∂x}:  \frac{∂b}{∂x}]$$

**Comment by albanD** [comment]
Hi,
Thanks for sharing this. I'm afraid this is expected behavior here. Similar to other issues such as https://github.com/pytorch/pytorch/issues/70342

The problem here is not the where itself but the pow operation.
In the first example, the pow happens after the where. So during the pow operation never sees NaN and there are not issues during the backward
In the second exmaple, the pow happens before the where. This means that, even ignoring the gradient generated by the where op, the bakcward of the pow function will do: `grad_input = 2 * input * grad_output`. And so, if input is NaN here, the grad_input will always be NaN (whatever grad_output coming from where contains).

One way to circumvent this would be to have a grad_output such that, when multiplied with NaN, you still get nothing. That's what masked Tensors are for: https://pytorch.org/docs/stable/masked.html?highlight=masked#module-torch.masked  They introduce the concept of "missing values" which will ensure that, even when multiplied by NaN, will remain "missing".

**Comment by randolf-scholz** [comment]
@albanD Yes that's definitely what's happening internally, and it seems like it is an inheret definiciency of how backward mode works -- forward mode computes the gradients correctly:

```python
import functorch
from torch import nn, tensor

x = nn.Parameter(tensor([1.1, 1.2, 1.3]))
y = tensor([1, 2, float("nan")])

def f(x):
    r = x - y
    r = r**2
    m = torch.isnan(y)
    r = torch.where(m, 0.0, r)
    return r.sum() / m.sum()

df = functorch.jacfwd(f)
print(df(x))   # [ 0.2000, -1.6000,  0.0000] ✔
df = functorch.jacrev(f)
print(df(x))   # [ 0.2000, -1.6000,     nan] ✘
```

When accumulating backwards, we do not know which branch the outer gradients came from, so in a sense what would need to happen is that when computing in reverse mode, torch would need to memorize when a `where` happened and then patch in  corresponding `where` operations in the following nodes (cf. the equations marked correct/incorrect in my previous post).

**Comment by albanD** [comment]
Well, just as this is dependent on the order, this order is indeed not problematic for forward mode. As the order of computation is the same, if you do the masking in the right order for the forward pass, it will do it in the right order for the forward mode gradients as well.

>  If masked tensors are how this should be solved in the end, then where-operations should possibly return masked gradients even if called with regular tensors.

This is going to be BC-breaking so will take a bit of time, but yes, this is the long term goal.

**Comment by randolf-scholz** [comment]
> Well, just as this is dependent on the order, this order is indeed not problematic for forward mode. As the order of computation is the same, if you do the masking in the right order for the forward pass, it will do it in the right order for the forward mode gradients as well.

You are hinting that forward mode can break as well due to bad order of operations. Do you happen to have any educational examples? Because I can't think of anything.

**Comment by soulitzer** [comment]
For forward mode, the gradients are computed in the same order that forward is computed, so that if the masking is done correctly, there should be no NaNs.
